sentry-core/src/metrics.rs

```diff
@@ -0,0 +1,404 @@
+
+enum Task {
+    SendMetrics((BucketKey, BucketValue)),
+    Flush,
+    Shutdown,
+}
+
+pub struct MetricFlusher {
+    sender: SyncSender<Task>,
+    shutdown: Arc<AtomicBool>,
+    handle: Option<JoinHandle<()>>,
+}
+
+const FLUSH_INTERVAL: Duration = Duration::from_secs(10);
+
+impl MetricFlusher {
+    pub fn new(transport: TransportArc) -> Self {
+        let (sender, receiver) = sync_channel(30);
+        let shutdown = Arc::new(AtomicBool::new(false));
+        let shutdown_worker = shutdown.clone();
+        let handle = thread::Builder::new()
+            .name("sentry-metrics".into())
+            .spawn(move || Self::worker_thread(receiver, shutdown_worker, transport))
+            .ok();
+
+        Self {
+            sender,
+            shutdown,
+            handle,
+        }
+    }
+
+    pub fn send_metric(&self, metric: &str) {
+        fn mk_value(ty: MetricType, value: &str) -> Option<BucketValue> {
+            Some(match ty {
+                MetricType::Counter => BucketValue::Counter(value.parse().ok()?),
+                MetricType::Distribution => BucketValue::distribution(value.parse().ok()?),
+                MetricType::Set => BucketValue::set_from_str(value),
+                MetricType::Gauge => BucketValue::gauge(value.parse().ok()?),
+            })
+        }
+
+        fn parse(metric: &str) -> Option<(BucketKey, BucketValue)> {
+            let mut components = metric.split('|');
+            let mut values = components.next()?.split(':');
+            let name = values.next()?;
+
+            let ty: MetricType = components.next().and_then(|s| s.parse().ok())?;
+            let mut value = mk_value(ty, values.next()?)?;
+
+            for value_s in values {
+                value.merge(mk_value(ty, value_s)?).ok()?;
+            }
+
+            let mut tags = "";
+            let mut timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap_or_default()
+                .as_secs();
+
+            for component in components {
+                if let Some(component_tags) = component.strip_prefix('#') {
+                    tags = component_tags;
+                } else if let Some(component_timestamp) = component.strip_prefix('T') {
+                    timestamp = component_timestamp.parse().ok()?;
+                }
+            }
+
+            Some((
+                BucketKey {
+                    timestamp,
+                    ty,
+                    name: name.into(),
+                    tags: tags.into(),
+                },
+                value,
+            ))
+        }
+
+        if let Some(parsed_metric) = parse(metric) {
+            let _ = self.sender.send(Task::SendMetrics(parsed_metric));
+        }
+    }
+
+    pub fn flush(&self) {
+        let _ = self.sender.send(Task::Flush);
+    }
+
+    fn worker_thread(receiver: Receiver<Task>, shutdown: Arc<AtomicBool>, transport: TransportArc) {
```

Comment:
Why does this have the `shutdown` parameter in addition to the `Shutdown` variant of `Task`? Wouldn't sending `Task::Shutdown` be enough?

Reply:
Thats a good point, though I copied most of this code from the transport that has the same logic. I can’t remember why that was, but also I don’t want to change it too much tbh.

Reply:
well, it looks like the tests were not happy with the way that explicit forceful shutdown and the channel interacted with each other, so I removed that.
The difference to the transport is that the transport can have long running http requests, and you can explicitly flush with a timeout, and then force-shutdown.